wolfcrypt/src/ed25519.c

```diff
@@ -388,18 +490,81 @@ int wc_ed25519ph_sign_msg(const byte* in, word32 inLen, byte* out,
 
 #ifdef HAVE_ED25519_VERIFY
 
+/*
+   sig        is array of bytes containing the signature
+   sigLen     is the length of sig byte array
+   key        Ed25519 public key
+   return     0 on success
+   type       variant to use -- Ed25519, Ed25519ctx, or Ed25519ph
+   context    extra signing data
+   contextLen length of extra signing data
+*/
+static int ed25519_verify_msg_init_with_sha(const byte* sig, word32 sigLen, ed25519_key* key, wc_Sha512 *sha,
+                               byte type, const byte* context, byte contextLen)
+{
+    int ret;
+
+    /* sanity check on arguments */
+    if (sig == NULL || key == NULL ||
+        (context == NULL && contextLen != 0)) {
+        return BAD_FUNC_ARG;
+    }
+
+    /* check on basics needed to verify signature */
+    if (sigLen != ED25519_SIG_SIZE || (sig[ED25519_SIG_SIZE-1] & 224))
+        return BAD_FUNC_ARG;
+
+    /* find H(R,A,M) and store it as h */
+
+#ifdef WOLFSSL_ED25519_PERSISTENT_SHA
+    ret = ed25519_hash_reset(key);
+    if (ret != 0)
+        return ret;
+#else
+    ret = 0;
+#endif
+
+    if (type == Ed25519ctx || type == Ed25519ph) {
+        ret = ed25519_hash_update(key, sha, ed25519Ctx, ED25519CTX_SIZE);
+        if (ret == 0)
+            ret = ed25519_hash_update(key, sha, &type, sizeof(type));
+        if (ret == 0)
+            ret = ed25519_hash_update(key, sha, &contextLen, sizeof(contextLen));
+        if (ret == 0 && context != NULL)
+            ret = ed25519_hash_update(key, sha, context, contextLen);
+    }
+    if (ret == 0)
+        ret = ed25519_hash_update(key, sha, sig, ED25519_SIG_SIZE/2);
+    if (ret == 0)
+        ret = ed25519_hash_update(key, sha, key->p, ED25519_PUB_KEY_SIZE);
+
+    return ret;
+}
+
+/*
+   msgSegment     an array of bytes containing a message segment
+   msgSegmentLen  length of msgSegment
+   key            Ed25519 public key
+   return         0 on success
+*/
+static int ed25519_verify_msg_update_with_sha(const byte* msgSegment, word32 msgSegmentLen,
+                                        ed25519_key* key, wc_Sha512 *sha) {
+    /* sanity check on arguments */
+    if (msgSegment == NULL || key == NULL)
+        return BAD_FUNC_ARG;
+
+    return ed25519_hash_update(key, sha, msgSegment, msgSegmentLen);
+}
+
 /*
    sig     is array of bytes containing the signature
    sigLen  is the length of sig byte array
-   msg     the array of bytes containing the message
-   msgLen  length of msg array
    res     will be 1 on successful verify and 0 on unsuccessful
    key     Ed25519 public key
    return  0 and res of 1 on success
 */
-int wc_ed25519_verify_msg_ex(const byte* sig, word32 sigLen, const byte* msg,
-                              word32 msgLen, int* res, ed25519_key* key,
-                              byte type, const byte* context, byte contextLen)
+static int ed25519_verify_msg_final_with_sha(const byte* sig, word32 sigLen, int* res,
```

Comment:
Nit pick... Can we fix the line length here.

Reply:
done